File: apps/haskell/src/Ch07/VoteCounter.hs
module Ch07.VoteCounter
    ( countVotes
    , countVotesParallel
    ) where

import Control.Concurrent.Async (mapConcurrently)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.List (foldl')

-- | Count votes sequentially
countVotes :: [String] -> Map String Int
countVotes = foldl' countOne Map.empty
  where
    countOne acc vote = Map.insertWith (+) vote 1 acc

-- | Count votes using Fork/Join pattern
countVotesParallel :: [String] -> IO (Map String Int)
countVotesParallel [] = return Map.empty
countVotesParallel votes = do
    -- Split into chunks and process in parallel
    let chunks = splitIntoChunks 4 votes
    results <- mapConcurrently (return . countVotes) chunks
    return $ foldl' (Map.unionWith (+)) Map.empty results

-- | Split a list into n chunks
splitIntoChunks :: Int -> [a] -> [[a]]
splitIntoChunks n xs =
    let len = length xs
        chunkSize = (len + n - 1) `div` n
    in go chunkSize xs
  where
    go _ [] = []
    go size ys = take size ys : go size (drop size ys)

File: apps/fsharp/src/Ch07/VoteCounter.fs
namespace GrokkingConcurrency.Ch07

open System

module VoteCounter =
    /// Count votes sequentially
    let countVotes (votes: string list) : Map<string, int> =
        votes
        |> List.groupBy id
        |> List.map (fun (key, group) -> (key, List.length group))
        |> Map.ofList

    /// Merge two vote count results
    let mergeResults (a: Map<string, int>) (b: Map<string, int>) : Map<string, int> =
        let allKeys = Set.union (Map.keys a |> Set.ofSeq) (Map.keys b |> Set.ofSeq)
        allKeys
        |> Set.toList
        |> List.map (fun key ->
            let countA = Map.tryFind key a |> Option.defaultValue 0
            let countB = Map.tryFind key b |> Option.defaultValue 0
            (key, countA + countB))
        |> Map.ofList

    /// Count votes using fork-join pattern
    let countVotesParallel (votes: string list) : Map<string, int> =
        if List.isEmpty votes then
            Map.empty
        else
            let numCores = Environment.ProcessorCount
            let chunkSize = max 1 (List.length votes / numCores)
            let chunks = votes |> List.chunkBySize chunkSize

            let results =
                chunks
                |> List.map (fun chunk -> async { return countVotes chunk })
                |> Async.Parallel
                |> Async.RunSynchronously
                |> Array.toList

            results |> List.reduce mergeResults
